src/persist/src/file.rs

```diff
@@ -182,6 +183,198 @@ impl Persister for FilePersister {
+    pub fn new<P: AsRef<Path>>(base_dir: P, lock_info: &[u8]) -> Result<Self, Error> {
+        let base_dir = base_dir.as_ref();
+        fs::create_dir_all(&base_dir)?;
+        {
+            let lockfile_path = Self::lockfile_path(&base_dir);
+            let mut lockfile = OpenOptions::new()
+                .write(true)
+                .create_new(true)
+                .open(lockfile_path)?;
+            lockfile.write_all(&lock_info)?;
+            lockfile.sync_all()?;
+        }
+        let buffer_path = Self::buffer_path(&base_dir);
+        let mut buffer_file = OpenOptions::new()
+            .append(true)
+            .read(true)
+            .create(true)
+            .open(&buffer_path)?;
+
+        let metadata_path = Self::metadata_path(&base_dir);
+        let mut metadata_file = OpenOptions::new()
+            .write(true)
+            .read(true)
+            .create(true)
+            .open(&metadata_path)?;
+
+        // Retrieve the last truncated sequence number from the metadata file
+        // unless that file was just created, in which case default to 0.
+        let mut bytes = Vec::new();
+        metadata_file.read_to_end(&mut bytes)?;
+        let seqno_start = if bytes.len() != 0 {
+            assert!(bytes.len() == 8);
+            let mut bytes = &bytes[..];
+            let mut buf = [0u8; 8];
+            // Decode sequence number
+            bytes.read_exact(&mut buf)?;
+            SeqNo(u64::from_le_bytes(buf))
+        } else {
+            metadata_file.write_all(&0u64.to_le_bytes())?;
+            metadata_file.sync_all()?;
+            SeqNo(0)
+        };
+
+        // Retrieve the most recently written sequence number in the buffer,
+        // unless the buffer is empty in which case default to 0.
+        let seqno_end = {
+            let len = buffer_file.metadata()?.len();
+
+            if len > 0 {
+                let mut buf = [0u8; 8];
+                buffer_file.seek(SeekFrom::End(8))?;
+                buffer_file.read_exact(&mut buf)?;
+                SeqNo(u64::from_le_bytes(buf))
+            } else {
+                SeqNo(0)
+            }
+        };
+
+        Ok(FileBuffer {
+            dataz: Arc::new(Mutex::new((buffer_file, metadata_file))),
+            seqno: seqno_start..seqno_end,
+            buf: Vec::new(),
+        })
+    }
+
+    fn lockfile_path(base_dir: &Path) -> PathBuf {
+        base_dir.join(Self::LOCKFILE_PATH)
+    }
+
+    fn buffer_path(base_dir: &Path) -> PathBuf {
+        base_dir.join(Self::BUFFER_PATH)
+    }
+
+    fn metadata_path(base_dir: &Path) -> PathBuf {
+        base_dir.join(Self::METADATA_PATH)
+    }
+}
+
+impl Buffer for FileBuffer {
+    fn write_sync(&mut self, buf: Vec<u8>) -> Result<SeqNo, Error> {
+        self.buf.clear();
+        let write_seqno = self.seqno.end;
+        self.seqno = self.seqno.start..SeqNo(self.seqno.end.0 + 1);
+        // Write length prefixed data, and then the sequence number.
```

Comment:
lets document the format in the rustdoc for FileBuffer. ditto for the metadata file format